state/protocol/badger/mutator.go

```diff
@@ -815,30 +809,17 @@ func (m *FollowerState) handleServiceEvents(block *flow.Block) ([]func(*transact
 
 			case *flow.EpochCommit:
 
-				// We should only have a single epoch commit event per epoch.
-				if epochStatus.NextEpoch.CommitID != flow.ZeroID {
-					// true iff EpochCommit event for NEXT epoch was already included before
-					return nil, state.NewInvalidExtensionError("duplicate epoch commit service event")
-				}
-
-				// The epoch setup event needs to happen before the commit.
-				if epochStatus.NextEpoch.SetupID == flow.ZeroID {
-					return nil, state.NewInvalidExtensionError("missing epoch setup for epoch commit")
-				}
-
-				// The commit event should have the counter increased by one.
-				if ev.Counter != counter+1 {
-					return nil, state.NewInvalidExtensionErrorf("next epoch commit has invalid counter (%d => %d)", counter, ev.Counter)
-				}
-
-				// Finally, the commit should commit all the necessary information.
-				setup, err := m.epoch.setups.ByID(epochStatus.NextEpoch.SetupID)
+				extendingSetup, err := m.epoch.setups.ByID(epochStatus.NextEpoch.SetupID)
 				if err != nil {
 					return nil, state.NewInvalidExtensionErrorf("could not retrieve next epoch setup: %s", err)
 				}
-				err = isValidEpochCommit(ev, setup)
-				if err != nil {
-					return nil, state.NewInvalidExtensionErrorf("invalid epoch commit: %s", err)
+				// validate the service event
+				err = isValidExtendingEpochCommit(ev, extendingSetup, activeSetup, epochStatus)
```

Comment:
This is checked when a block is finalized, if it's invalid extension we still have to accept it because this block has been finalized. But storing a finalized block with invalid service event is weird to me.

Why no checking when a block is incorporated and reject if is invalid? So that every finalized block must have passed that check

Reply:
Here's how I think about it: It's tricky because an invalid service event is a bit different than an invalid message from another node. It's different because the originator of the message (service account) is trusted.

If we see an invalid service event, either: (1) there is a bug in the service account, or (2) the service event was maliciously modified or injected by a malicious execution node.

If (1), then all honest ENs and VNs will agree on the state resulting from emission of the invalid service event, and that state will be sealed. If we discard receipts and approvals that attest to this state, sealing will halt. So we need to incorporate the service event anyway if we want to preserve liveness. 

If (2), then the service event won't be sealed because verification of the result its part of will fail. So it is safe to incorporate the malicious result. We don't _apply_ state changes resulting from service events until after they are sealed for exactly this reason.